x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/loadingservice/ModelLoadingService.java

```diff
@@ -382,19 +379,22 @@ private void handleLoadSuccess(String modelId,
             trainedModelCircuitBreaker);
         synchronized (loadingListeners) {
             listeners = loadingListeners.remove(modelId);
-            // If there is no loadingListener that means the loading was canceled and the listener was already notified as such
-            // Consequently, we should not store the retrieved model
+            // if there are no listeners, simply release and leave
             if (listeners == null) {
                 loadedModel.release();
                 return;
             }
 
-            // temporarily increase the reference count before adding to
-            // the cache in case the model is evicted before the listeners
-            // are called in which case acquire() would throw.
-            loadedModel.acquire();
-            localModelCache.put(modelId, new ModelAndConsumer(loadedModel, consumer));
-            shouldNotAudit.remove(modelId);
+            // If the model is referenced, that means it is currently in a pipeline somewhere
+            // Also, if the consume is a search consumer, we should always cache it
+            if (referencedModels.contains(modelId) || consumer.equals(Consumer.SEARCH)) {
```

Comment:
lines 403-404 call release on localModel I think those lines should be wrapped in the same conditional or move the acquire outside the if

```  
L403   // account for the acquire in the synchronized block above
L404   loadedModel.release();
```